internals/overlord/logstate/loki/loki_test.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki_test
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/canonical/pebble/internals/overlord/logstate/loki"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+type suite struct{}
+
+var _ = Suite(&suite{})
+
+func Test(t *testing.T) {
+	TestingT(t)
+}
+
+func (*suite) TestRequest(c *C) {
+	input := []servicelog.Entry{{
+		Time:    time.Date(2023, 12, 31, 12, 34, 50, 0, time.UTC),
+		Service: "svc1",
+		Message: "log line #1\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 51, 0, time.UTC),
+		Service: "svc2",
+		Message: "log line #2\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 52, 0, time.UTC),
+		Service: "svc1",
+		Message: "log line #3\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 53, 0, time.UTC),
+		Service: "svc3",
+		Message: "log line #4\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 54, 0, time.UTC),
+		Service: "svc1",
+		Message: "log line #5\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 55, 0, time.UTC),
+		Service: "svc3",
+		Message: "log line #6\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 56, 0, time.UTC),
+		Service: "svc2",
+		Message: "log line #7\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 57, 0, time.UTC),
+		Service: "svc1",
+		Message: "log line #8\n",
+	}, {
+		Time:    time.Date(2023, 12, 31, 12, 34, 58, 0, time.UTC),
+		Service: "svc4",
+		Message: "log line #9\n",
+	}}
+
+	expected := `
+{
+  "streams": [
+    {
+      "stream": {
+        "pebble_service": "svc1"
+      },
+      "values": [
+          [ "1704026090000000000", "log line #1" ],
+          [ "1704026092000000000", "log line #3" ],
+          [ "1704026094000000000", "log line #5" ],
+          [ "1704026097000000000", "log line #8" ]
+      ]
+    },
+    {
```

Comment:
Let's please use the same convention used above for the Go values to avoid the excessive spacing here.